add PR template

pdpinch opened the pull request:
<!--
Please give the pull request a short but descriptive title.
Use [conventional commits](https://www.conventionalcommits.org/) to separate and summarize commits logically.

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/edx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

## Description

This PR adds a pull request template in Github. The contents of the template will be used as the default content of a pull request description, and PR authors will edit it before submitting a pull request. 

This PR will only affect developers. 

Screenshot of pull request template as the PR author will see it:

<img width="933" alt="image" src="https://user-images.githubusercontent.com/430126/105848376-e0796d80-5fac-11eb-8394-82d23af111a9.png">

Screenshot of pull request template as reviewer will see it:

<img width="746" alt="image" src="https://user-images.githubusercontent.com/430126/105848420-f25b1080-5fac-11eb-83f6-1194c7a9960a.png">

## Supporting information

Fixes https://github.com/edx/open-edx-proposals/issues/171

## Testing instructions

As far as I know, the only way I know to test PR templates is to merge them and create a PR

1. If you don't already have a suitable fork, fork edx-platform 
2. Merge this PR into master on your fork
3. Open a PR _on your fork_ 
4. The PR description should be pre-populated with the contents of `pull_request_template.md`

## Deadline

None

## Other information

The template includes a link to an OEP PR. Once that OEP has been merged, the template should be updated.

edx-webhook commented:
Thanks for the pull request, @pdpinch! I've created [OSPR-5315](https://openedx.atlassian.net/browse/OSPR-5315) to keep track of it in JIRA.

As a core committer in this repo, you can merge this once the pull request is approved per the [core committer reviewer requirements](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) and according to the agreement with your edX Champion.



<!-- comment:welcome-core-committer -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

nasthagiri commented:
@pdpinch Thanks for putting this in a PR. Can you add the "meta" (desired outcomes of each element) as [markdown comments](https://www.jamestharpe.com/markdown-comments/) in the template?  This will allow us to co-locate the rationale for what we choose to include and exclude - so we can reassess overtime.

natabene commented:
@pdpinch Thank you for your contribution. @nedbat You will want to have a look at this. 
My personal wish, it is possible to do - include information on which role the change will have impact. E.g. learner, course author, developer, etc. That would help me quite a bit. I have to ensure all PR is with UI or behavior changes, visible to learners, course author, or course admins go through a product review first. Indication of role impacted would help me streamline this process. But if it adds to much burden on PR author, it is fine.

nedbat commented:
I love the idea of including reasoning for the information we are requesting.  But markdown comments are not a good way to do it unless you want people editing their pull requests to see them.  (BTW, that link about markdown comments is very odd: most of the syntax doesn't work as comments on GitHub at all, and three dashes are unnecessary).

Maybe an ADR (or OEP since this will likely be rolled out more broadly) for the rationale?

nasthagiri commented:
An ADR is fine as well.  I mentioned MD comments since (1) we don't want to display the rationale in the PR template itself (the rationale is needed for long-term maintenance of the template, not for completing the template), and (2) it's the closest location to the template elements.  I'd like to be able to read the template content, annotated with why it's there.

If Github allows the template to be in RST format, we can use RST comments instead.

nasthagiri commented:
@natabene I like the idea of prompting OSPR contributors to complete a short-survey that will help us route the PR faster.  You mention providing information on whether the PR impacts UI and which user persona (author/learner/etc). That's a great idea. Can we put this survey in the OSPR comment that is added _after_ the BOT determines that it's an OSPR - rather than in the default PR Template for all PRs?  _(I'm trying to ensure the PR template is relevant to the vast majority of our PRs - so it doesn't get ignored over time.)_

nedbat commented:
It doesn't matter what markup we use (rst or md): the author of the PR will be presented with the full text including comments, when creating a pull request.  If we have text like rationales that we don't want to distract the author of the pull request, then it has to go somewhere other than the pull request template.

About UI impact and personas affected: this seems like useful information for all pull requests, especially as we are trying to tighten up how we communicate changes to the code.

pdpinch commented:
Thinking about where to put rationales led me to take a look at [CONTRIBUTING.rst](https://github.com/edx/edx-platform/commit/0717e28e21cf8271193ff80acb48cdcd314a293f#diff-76e7ca2f2843ac5f659b45382768c1b64bf8704e8d3c6189c4e3623f5d465bc2) which lead me eventually to https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/process/cover-letter.html -- which I probably should have read before I got started. 

Clearly the developer's guide needs an update. Would that be a good place to link to for rationales?

pdpinch commented:
@natabene I added a section for "user impact" but I'm a little torn because I feel like this should naturally be part of the description. On the other hand, it would be valuable for me as a reviewer to see at glance which users are impacted.

nedbat commented:
It feels like we have a few different approaches to the same problem in flight.  I took the liberty of trying to collect them together: https://github.com/edx/open-edx-proposals/pull/180  This could be the place to put the rationale for the pull request template.

TBH, it was hard to come up with reasons for why it's good to provide this information, since it seems kind of self-evident.  I'm happy to collaborate on them.

nasthagiri commented:
> it was hard to come up with reasons for why it's good to provide this information, since it seems kind of self-evident. 

@nedbat Can you expand further on what you mean by this statement? Is this in regards to the rationale for the requested information in the PR template?

nedbat commented:
>> it was hard to come up with reasons for why it's good to provide this information, since it seems kind of self-evident.
>
> @nedbat Can you expand further on what you mean by this statement? Is this in regards to the rationale for the requested information in the PR template?

Yes, sorry, this is about the rationales.  As I was trying to express why we wanted the information, I was having a hard time coming up with detailed reasons because it seems obvious to me why this information is useful.

nasthagiri commented:
> it seems obvious to me why this information is useful
Yes, I can see that. For those immersed in the project, the reasons may seem obvious - hence even more so important to externalize them for others, IMO. Also, when we externalize the outcomes, we may find duplicative efforts that we can optimize.

> So I could use some help.
I'm a bit confused about what and where we want to document at this point. This particular PR and its parent GitHub Issue are specific to PR templates (and also specific to `edx-platform` right now). I think a PR template is a good starting point and we have concrete ideas for it right now - so we can move forward on it.

nedbat commented:
Absolutely we can move forward with the PR template, I didn't mean to suggest otherwise.  I liked your idea of capturing the motivation for collecting information. Since we can't do it in the template, I started another document.

nedbat commented:
I tightened up the text, and reduced the number of sections, so that it didn't feel like a multi-part form with required sections.  What do you think?:

```
<!-- 
Use this template as a guide. Omit sections that don't apply.
More details about the template are at https://github.com/edx/open-edx-proposals/pull/180 
-->

## Description

Describe what this pull request changes, and why. Include implications for people using this change.
Design decisions should be documented in the repo (docstring / README / ADR) and linked here.

Useful information:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
"Developer" and "Admin".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Configuration changes needed.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions. 
Be sure to check they are publicly readable, or if not, repeat the information here.

## Deadline

"None" if there's no rush, "ASAP" if it's critical, or provide a specific date or event if there is one.

## Other information

Include anything that will help reviewers and consumers of the change.
- Does this change depend on other changes elsewhere?
- How can it be tested?
- Any special concerns or limitations? Migrations? A11y?
```

pdpinch commented:
Thanks @nedbat. My only hesitation is the diminished importance of testing instructions. I consider those the most useful part of the PR, aside from the description. What do you think of bumping "How can it be tested" up to the 1st or 2nd bullet under "useful information" ?

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+<!--
+Use this template as a guide. Omit sections that don't apply.
+More details about the template are at https://github.com/edx/open-edx-proposals/pull/180
```
Can we have the rationale and expected outcomes of the template documented somewhere in the repo itself? Perhaps in a top-level ADR folder.

nasthagiri replied:
Relatedly, if one were to measure the impact of having this template, what would the hypotheses be?  I want to ensure we are aligned on the value of this effort (rollout, adopt, future maintenance).

nasthagiri replied:
Idea: You can start with an ADR in `edx-platform` for now, since that's where we're rolling this out - and we need a place document this decision. Each repo that chooses to adopt this template, can have their own ADR for the adoption and link to the one in edx-platform. The repo may have their own modifications/additions, which they can capture in their repo-specific ADR.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+
+Useful information:
+- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
+"Developer" and "Admin".
```
What does "Admin" mean in this context? "Course Admin" and "Django Admin" are very different personas. Maybe include "Operator"?

Also, do we want this to be a structured field that we can parse for and use in automation? Or is this a reminder for folks to include in their description?

Finally, in order to "sell" and "maintain" this info, it would be great to know _why_ we believe this info is necessary to ask people to specify.

nedbat replied:
I've changed it to "Operator".

As I've said, I could use some help expressing the "why" for each piece of information. I would rather not put all the whys into the PR template: they will make it bulkier and harder to navigate.  We have a draft of a document for the reasons and backstory for the information we need from developers: https://github.com/edx/open-edx-proposals/pull/180  Can we add to that document?

nasthagiri replied:
Indeed. Scratch my earlier comment about putting the rationale in an ADR.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+Useful information:
+- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
+"Developer" and "Admin".
+- How can it be tested?
```
"How can it be tested" seems out of place in the "Description" section.

pdpinch replied:
As I noted earlier, I feel that "how can it be tested" is one of the most useful parts of the PR description. I fear that if it is relegated to "Other Information" in a middling bullet it will be omitted regularly. 

If it doesn't belong here, can we elevate it to the ## level?

nasthagiri replied:
@pdpinch Can you clarify your position with a reference example of what information you expect from the "how can it be tested" section?  Also, what outcome are you expecting? Faster review cycle? Is the test instructions intended to be short-term just for the review cycle or longer-term for later reference?

pdpinch replied:
I wrote a "Testing instructions" section in the [PR description for this PR](https://github.com/edx/edx-platform/pull/25949#issue-546091511). 

Outcome: "Testing instructions" are primarily intended to speed up the review cycle, avoiding the reviewer wasting time trying to infer the steps to test from the description. Instructions can also be useful for later reference if a future developer is researching regression.

sarina replied:
I love testing instructions in PRs. It shows the author has thought carefully about the change and gives me a place to start.

sarina replied:
I do wonder if `How can it be tested` is the best wording. My suggestion might be, `Please provide step-by-step test instructions` (to be explicit that we want this, and what we are looking for)

nedbat replied:
I've moved testing instructions to a top-level section.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+"Developer" and "Admin".
+- How can it be tested?
+- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
+- Configuration changes needed.
```
Instead of having people include config info in the PR description, what if we remind them to annotate their configs and link to the config annotations?  Robert may have a suggestion for this.

nedbat replied:
"Provide links to the description of corresponding configuration changes. Remember to correctly annotate these changes."

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
+- Configuration changes needed.
+
+## Supporting information
```
Nice. I like how it asks folks to ensure public readability and copy the info, if not.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+
+## Description
+
+Describe what this pull request changes, and why. Include implications for people using this change.
```
If we ask folks to document and duplicate the "why" in the PR description, how do we ensure and clarify that the "why" needs to be documented in docstrings and ADRs?  Instead, should we nudge them to document in ADRs and ask them to link to the "why" in the description?

nedbat replied:
The PR description should have the why of the PR as a whole.  ADRs document design decisions made along the way. Most PRs don't have ADRs.  Docstrings should be focused on the specific function/class/module that they document. I don't think there will be much duplication among these three different places.

nasthagiri replied:
ADRs document both the decisions and the rationale behind those decisions.

To rephrase my concern: I'd like to remind and encourage folks to adopt ADRs as part of their work. The ADRs provide the historical trace of the changes and their rationale. The adoption of the ADR process is what allows us to not have a central blocking architectural council. Thus, I want to ensure folks do not consider adding docs in PR descriptions as a replacement for ADRs.

nedbat replied:
Do we agree that most PRs will not involve an ADR and that most PRs will want a "why" described?  I wonder if we are starting from different understandings.  I definitely agree that ADR-worthy decisions should be recorded in ADRs, and that there's no need to duplicate that decision process in the PR template.

nasthagiri replied:
I agree that many PRs do not need an ADR.
I believe the wording below can be adjusted to address the "nudge" that I'm looking for:

`Reminder that design decisions and their rationales are to be documented in the repo (docstring / ADR), per OEP-19, and can be linked here.`

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+
+## Description
+
+Describe what this pull request changes, and why. Include implications for people using this change.
```
Also, is this a good opportunity to [link to OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html#adrs), which discusses what should go in READMEs versus ADRs versus comments.

pdpinch replied:
I added a link per @nedbat's suggestion.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+
+Include anything that will help reviewers and consumers of the change.
+- Does this change depend on other changes elsewhere?
+- Any special concerns or limitations? Migrations? A11y?
```
Seems like info about testing could probably go in this section.

sarina replied:
Should we spell out the word `a11y` for developers who may be new to the concept of accessibility?

sarina replied:
@nasthagiri I'm not sure I agree that `Testing` should be nested under `Other Information`. I think testing is very important to have thought through and spell out that having it as a higher-up section makes sense to me.

nasthagiri replied:
I'm supportive of us including "Testing" in its own section, assuming we agree on its value.  "Testing" didn't seem to belong in the "Description" section.  @pdpinch has taken a stab at proposing its value in [this comment](https://github.com/edx/edx-platform/pull/25949#discussion_r561082418).

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+
+## Deadline
+
+"None" if there's no rush, "ASAP" if it's critical, or provide a specific date or event if there is one.
```
@natabene Can you take a look at this "Deadline" section? Are there other field values for "deadline" that would be useful for OSPR triage?

Also, "ASAP" seems too general. Does that mean minutes/hours/days?

nedbat replied:
I've removed ASAP.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+<!--
```
Should any of these be checklists instead of bullet points?

nedbat replied:
I didn't see any of them as simple check-offs, since the instructions are asking for information. Perhaps, "[ ] I have written a docstring/readme/adr" could be a check mark, but I'd rather have a link to it.

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,31 @@
+<!--
```
Where do you plan to include a check for PR description and commit message?

nedbat replied:
I'm not sure what you mean by a check? An automated check?

nasthagiri replied:
I meant a checkbox. But any other solution is also fine - as a reminder.

pdpinch replied:
I'm not clear what a reminder for commit messages might say. I thought the commit message linting was going to be checked by some kind of automation?

nedbat replied:
For PR title and commits, how about:
```
Please give the pull request a short but descriptive title.
Use [conventional commits](https://www.conventionalcommits.org/) to separate and summarize commits logically.
```
@nasthagiri you said something today about not using Jira issue numbers in the title.  Some of our existing automation relies on this, and people find the short-hand useful, so I'm not sure we should push people away from that habit.

nasthagiri replied:
I was under the impression that our existing automation would look at the JIRA numbers in the PR description in addition to the title.  If there's a risk of breaking those, then ok, we can add that later.

nasthagiri replied:
+1 to the proposed change for the descriptive title and the conventional commits.

nasthagiri reviewed (commented):
Thanks for pulling this together. I like the overall structure. I have a few comments for considersation, which hopefully strengthen the proposal and its adoption.

nedbat commented:
BTW, some other possible things to cover, reading over [Change Transparency](https://github.com/edx/open-edx-proposals/pull/180): deprecations, reverts, and chores.

nedbat commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,34 @@
+## Description
+
+Describe what this pull request changes, and why. Include implications for people using this change.
+Design decisions should be documented in the repo (docstring / README / ADR) and linked here.
```
```suggestion
Design decisions and their rationales should be be documented in the repo (docstring / ADR), per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be linked here.
```

nedbat replied:
```suggestion
Design decisions and their rationales should be be documented in the repo (docstring / ADR), per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be linked here.
```

nedbat commented:
@pdpinch I've made one last tweak using Nimisha's words.  Are we good to go?

felipemontoya commented:
fwiw I left my general approval as a core commiter here. I don't have specific feedback, I very much liked what I saw on the template

nasthagiri commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,34 @@
+## Description
+
+Describe what this pull request changes, and why. Include implications for people using this change.
+Design decisions and their rationales should be be documented in the repo (docstring / ADR), per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be linked here.
```
Nit: `be be` -> `be`

nasthagiri reviewed (commented):
This looks good to me overall! Thanks for all the effort.

Let's include the final touches before merging.

pdpinch commented:
@nasthagiri @nedbat I made the requested changes, and updated the PR description to model the template. 

I also updated the commit message, but I'm new to conventional commits and this is an unusual one. Let me know what you think.

nedbat commented:
Conventional commits are one of those categorization exercises that can be tricky on outliers like this commit.  I might have used "chore", but I think "docs" is better: these words are explaining something to devs, so docs is a good fit.

nedbat commented:
Thanks @pdpinch and @nasthagiri :)

edx-webhook commented:
@ormsbee, @nasthagiri: thought you might like to know that pdpinch merged this pull request.

<!-- comment:champion_merge_ping -->

edx-webhook commented:
@pdpinch 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/25949&entry.1484655318=https://github.com/pdpinch&entry.752974735=2020-12-28+13:21&entry.1917517419=2021-01-27+17:39&entry.2133058324=Yes) so we can improve your experience in the future.
